### DISCUSSION TITLE
Please add an IComparer<T> argument to CollectionAttribute.ctor()

### DISCUSSION THREAD

**Discussion by GennadyMaltsev** [xunit/xunit#2928] [Ideas]
Currently, Collections can be sorted only by name because CollectionAttribute ctor has no other arguments.
I ask you to add a new constructor for it which will have both name and user-defined IComparer<T> arguments (where T is the test class to which CollectionAttribute is applied) to improve the ability to sort collections in ITestCollectionOrderer implementations. 
And maybe it would be nice to add a public property CollectionAttribute.Name to make it possible to access this Name in OrderTestCollections() method.
Thanks in advance!

**Comment by bradwilson** [comment]
Well, I see a few problems.

First, the things that can be in constructor arguments is very limited. You could put a type there (that would presumably be required to implement `IComparer<T>`), but not an instance of an `IComparer<T>`.

Second, how would you resolve the situation where multiple CollectionAttribute instances reference to different comparers? Which comparer wins?

Third, I don't honestly see how this is simpler than just implementing `ITestCollectionOrderer`. Can you provide a concrete example of usage that would illustrate _why_ this described feature is better than the existing feature?

**Comment by GennadyMaltsev** [comment]
Hello, thank you for an answer.
1. Instance is not required here, the type is enough. Because we can add any other attributes to the class and its methods/properties an use them to sort classes in desired order.
2. Maybe, we can allow only one implementation of IComparer<T> for same-named collections and throw an exception on each different one? I.e., [CollectionAttribute("Name1", MyComparer1)] can be adjacent to another [CollectionAttribute("Name1", MyComparer1)] or CollectionAttribute("Name1"), or CollectionAttribute("Name2", MyComparer2), or CollectionAttribute("Name3", MyComparer1), but cannot exist together with CollectionAttribute("Name1", MyComparer2) ?
3. My case: I have many test classes with methods that are declared with [TestPriority(xxxx)] attribute. 
And 1st class has methods with priorities 0, 10, 20, 30...70. 2nd class's methods have priorities 100, 110, 120...150. 3rd class's methods: 300, 310, 320. And so on, the latest class's have priorities 2500+. All methods of the each class are sorted in ITestCaseOrderer.OrderTestCases() implememtation.
But I want to additionally sort these **classes** not by the name but by their minimal priority. For this, I can put them into different collections on into the same one, if necessary.

The issue is - ITestCaseOrderer can sort only methods within ONE class and cannot sort all test methods together. And ITestCollectionOrderer cannot sort classes but can only sort collections by their names.

My test cases must be ordered by classes, and some classes must be run first. I think it's not too smart idea to add to each class (of 25 ones at the moment) a CollectionAttribute with alphabetically sorted abracadabras as the names. 
All of my test classes are derived from the TestBase class with common CollectionAttribute["NonParallelUnitTests"] and to several classes I added different-named CollectionAttribute to sort them using ITestCollectionOrderer implememtation.
But I want to use more beauty solution :) 

Possibly, you can suggest it? For example, add IComparer<T> in CollectionAttribute ctor(), as I already said, or add any method which can be invoked BEFORE launching of OrderTestCollections() which can forcibly change CollectionAttributes on the used classes, or add new CollectionAttribute, or add new custom priority attribute to each class and put it into different collection, or add new interface that can be implemented, like as I**TestClass**Orderer, or something similar.
In the end, the ability to sort classes in any case will not be an unnecessary option.
Thank you very much!

**Reply by bradwilson** [reply]
> My test cases must be ordered by classes, and some classes must be run first.

This sounds like what you really need is a custom framework implementation. Your needs are more specialized and complex than we can solve with a this feature request (which is itself quite complex).

Our sample custom framework: https://github.com/xunit/samples.xunit/tree/main/v3/ObservationExample

**Comment by olstakh** [comment]
I'd love to have `ITestClassOrderer` too. Currently classes within a collection are ordered by their `UniqueID`, which is outside of our control. Since we can sort collections (that may have multiple classes) and can sort test cases within one class - would be great to bridge the gap of sorting classes within a collection

**Reply by bradwilson** [reply]
Please open a new issue for this. Thanks!

**Reply by olstakh** [reply]
https://github.com/xunit/xunit/issues/3424
